heirloom/rm/rm.1
.\"
.\" Sccsid @(#)rm.1	1.11 (gritter) 12/6/04
.\" Parts taken from rm(1), Unix 7th edition:
.\" Copyright(C) Caldera International Inc. 2001-2002. All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"   Redistributions of source code and documentation must retain the
.\"    above copyright notice, this list of conditions and the following
.\"    disclaimer.
.\"   Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"   All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"      This product includes software developed or owned by Caldera
.\"      International, Inc.
.\"   Neither the name of Caldera International, Inc. nor the names of
.\"    other contributors may be used to endorse or promote products
.\"    derived from this software without specific prior written permission.
.\"
.\" USE OF THE SOFTWARE PROVIDED FOR UNDER THIS LICENSE BY CALDERA
.\" INTERNATIONAL, INC. AND CONTRIBUTORS ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL CALDERA INTERNATIONAL, INC. BE
.\" LIABLE FOR ANY DIRECT, INDIRECT INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
.\" BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
.\" WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
.\" OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
.\" EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.TH RM 1 "12/6/04" "Heirloom Toolchest" "User Commands"
.SH NAME
rm \- remove directory entries
.SH SYNOPSIS
\fBrm\fR [\fB\-f\&iRr\fR] \fIfile\fR ...
.SH DESCRIPTION
Rm removes the entries
for one or more files from a directory.
If an entry was the last link to the file,
the file is destroyed.
Removal of a file requires write permission in its directory,
but neither read nor write permission on the file itself.
.PP
If a file has no write permission
and the standard input is a terminal,
its permissions are printed
and a line is read from the standard input.
If that line begins with `y'
the file is deleted,
otherwise the file remains.
.PP
When
.B rm
encounters a symbolic link,
the link will be removed,
but its target will remain in the file hierarchy.
.PP
The
.B rm
command accepts the following options:
.TP
.B \-f
No questions are asked
and neither error messages are printed
nor the exit status is affected
in case of nonexistent files.
.B /usr/5bin/rm
and
.B /usr/5bin/s42/rm
will not print an error message
if removal of a file failed.
For
.BR /usr/5bin/posix/rm ,
any previous occurrences of the
.B \-i
option are ignored.
.TP
.B \-i
Causes
.B rm
to ask for confirmation
before deleting any file,
and,
unter
.BR \-r ,
whether to examine each directory.
For
.BR /usr/5bin/posix/rm ,
any previous occurrences of the
.B \-f
option are ignored.
.TP
.B \-R
Same as
.BR \-r .
This option has been introduced by POSIX.2.
.TP
.B \-r
.B Rm
will recursively delete the entire contents
of directories given as operands,
and the directory itself.
Without this option,
an error comment is printed
if a designated file is a directory.
.SH "ENVIRONMENT VARIABLES"
.TP
.B SYSV3
Causes the text of some messages to be changed.
.SH "SEE ALSO"
rmdir(1),
unlink(2),
rmdir(2)
.SH NOTES
It is forbidden to remove the files `.' and `..'
merely to avoid the antisocial consequences
of inadvertently doing something like `rm -r .*'.


ex-vi/Changes
Release ...
* When a multicolumn character was replaced by another multicolumn character
  in insert mode, the display was not updated appropriately with terminals
  other than xterm.
* Various rendering errors happened with multicolumn characters if they
  started at an even column (counting from 1 upwards).
* When a multicolumn character was inserted and then replaced, the visual
  screen representation was sometimes not updated accordingly.
* Undoing the replacement of a multicolumn character by a singlecolumn
  character twice made the singlecolumn character invisible.
* The 'cw' visual command with a multibyte character as last character of
  the affected word located at the end of the line left garbage bytes past
  the end of the new text.
* Visual 'U' followed by 'u' lead to garbage on the screen when multibyte
  characters were present on the changed line.
* The handling of multicolumn characters was fixed for terminals without the
  IM (insert mode) capability. It is unlikely that such terminals actually
  exist, but vi will use the corresponding code for open mode if a termcap
  entry is not available.
* When an illegal multibyte sequence is entered in vi insert mode, no garbage
  bytes are generated anymore when the insert mode is left.
* The '\u', '\l', '\U', and '\L' substitution sequences work with multibyte
  characters now.
* Handle character case conversions with the '~' vi command correctly if the
  length of the converted multibyte sequence is smaller than the original one.
* Multibyte sequences that correspond to an unprintable character are now
  printed as multiple octal escape sequences.
* Fixed an old vi bug: If a vi command that yanked or deleted part of a line
  was followed by an ex command that also yanked or deleted some text, a
  following 'p' vi command pasted the text affected by the former vi command.
  It now pastes the text of the last yank or delete even if that was an ex
  command.

Release 12/2/04
* Support for multibyte character locales was added.
* The code has been converted to ANSI C, and support for pre-POSIX systems has
  been dropped.
* When the end of the current line consists of blank characters and the 'w'
  visual command is given at this point, vi now advances to the start of the
  next line instead of ringing the bell. This is compatible with SVR4 vi and
  seems to be what POSIX specifies.
* If the replacement part of a substitute command consists of a single '%',
  as in ':s/foo/%/', the replacement part of the previous substitution is
  used. This is compatible with SVR4 vi and is specified by POSIX.
* Fixed a number of possible heap overflows, e.g. because of too long tag
  strings.

Release 6/5/04
* Some changes were made to the Makefile to support RPM builds. In particular,
  the meaning of the DESTDIR and PREFIX variables was changed.
* An insufficient size of a variable caused the window size set to 8 on
  FreeBSD if the terminal baud rate was 38400.

Release 1/3/04
* Changes to the included libterm only: Made multiple tc= capabilities in
  a termcap entry work; recognize tc= also if it is not the last capability
  in an entry (Bugreport by Andrew Minter).

Release 9/3/03
* The code did not check st_blksize from stat(2) at other points besides
  the one fixed in the last release.
* The keyboard input character with code 255 ("y in ISO-8859-1) was
  misinterpreted as end-of-file indicator.

Release 8/27/03
* Compile fixes for AIX and HP-UX (Mike Jetzer).
* Delete temporary file when preserving was successful after receiving
  SIGHUP or SIGTERM (Fix taken from 4.3BSD Reno).
* Set MAILRC to /dev/null in expreserve to avoid reading the user's
  mail initialization file.
* Optionally use Caldera's 'UNIX(R) Regular Expression Library' to
  get POSIX.2 compatible REs.
* Don't refuse to quit with 'No write since last change' if a line of a
  newly read input file is too long but no changes were made by the user.
* The POSIX_2 preprocessor define has been removed. The behavior previously
  enabled by this variable is now the default (except as noted below).
* Backslash inside RE bracket expresssions (as in [\]]) is now recognized
  as an escape character by default. This is not POSIX.2 compliant, but is
  compatible with historic vi behavior, is demanded by the some traditional
  standards such as the System V Interface definition, and is compatible
  with vim. To get POSIX.2 compliant behavior here, define NO_BE_BACKSLASH.
* The input buffer did overflow with large values of st_blksize from stat(2).

Release 4/3/02
* Option "showmode" has no effect on hardcopy or glass terminals (P).
* Fixed undo for :r! command.
* Interrupt character is not misinterpreted as DEL on insertion (P).
* Fixed interaction of <count>~ vi command with abbreviations and macros (P).
* Avoid horizontal scrolling of showmode string on wraparound to last line (P).
* No showmode string is printed when executing a macro.
* Recovery listing fixed not to mix characters from long filenames in
  /var/preserve with /var/tmp listing (P).
* Catch SIGXFSZ (file size limit exceeded).

Release 3/3/02
* Separated terse/noterse messages for RE errors (P).
* The expreserve and exrecover commands work again (P).
* Passing beyond EOL in <count>~ vi command is not an error (P).
* Fixed segmentation violation on mdjc'd and related bugs (Sven Mascheck).
* Marks remain on lines changed in an undo operation.
* Close mapmalloc file descriptor for /dev/zero on exec() (P).
* Added -L and -V command line options as in SVr4 ex.
* POSIX.2: Processing EXINIT does not depend on stdin being a terminal.
* POSIX.2: No newline is appended to an empty file in visual mode.
* Fixed segmentation violation on :f^V^M^M and similar commands.
* Mapmalloc extended to allocate and release multiple memory pools.
* Exrecover lists /var/tmp in addition to /var/preserve.
* Have multiple attempts to create named buffer temporary file.
* Size limit for temporary file names removed.

Release 2/17/02

heirloom/rm/rm.c
/*
 * rm - remove directory entries
 *
 * Gunnar Ritter, Freiburg i. Br., Germany, July 2002.
 */
/*
 * Copyright (c) 2003 Gunnar Ritter
 *
 * This software is provided 'as-is', without any express or implied
 * warranty. In no event will the authors be held liable for any damages
 * arising from the use of this software.
 *
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute
 * it freely, subject to the following restrictions:
 *
 * 1. The origin of this software must not be misrepresented; you must not
 *    claim that you wrote the original software. If you use this software
 *    in a product, an acknowledgment in the product documentation would be
 *    appreciated but is not required.
 *
 * 2. Altered source versions must be plainly marked as such, and must not be
 *    misrepresented as being the original software.
 *
 * 3. This notice may not be removed or altered from any source distribution.
 */

#if __GNUC__ >= 3 && __GNUC_MINOR__ >= 4
#define	USED	__attribute__ ((used))
#elif defined __GNUC__
#define	USED	__attribute__ ((unused))
#else
#define	USED
#endif
#ifdef	SUS
static const char sccsid[] USED = "@(#)rm_sus.sl	2.19 (gritter) 12/18/04";
#else
static const char sccsid[] USED = "@(#)rm.sl	2.19 (gritter) 12/18/04";
#endif

#include	<sys/types.h>
#include	<sys/stat.h>
#include	<sys/wait.h>
#include	<sys/resource.h>
#include	<fcntl.h>
#include	<unistd.h>
#include	<stdio.h>
#include	<string.h>
#include	<stdlib.h>
#include	<errno.h>
#include	<libgen.h>
#include	<dirent.h>
#include	<limits.h>
#include	<signal.h>
#include	<stdarg.h>

#include	"getdir.h"

struct	level {
	struct getdb	*l_db;
	int	l_fd;
};

static struct level	*levels;	/* visited directories */
static size_t	maxlevel;		/* maximum level visited */
static unsigned	errcnt;			/* count of errors */
static int	ontty;			/* stdin is a tty */
static int	fflag;			/* force */
static int	iflag;			/* ask for confirmation */
static int	rflag;			/* recursive */
static char	*progname;		/* argv[0] to main() */
static char	*path;			/* full path to current file */
static size_t	pathsz;			/* allocated size of path */
extern int	sysv3;			/* emulate SYSV3 behavior */

static int	subproc(size_t, const char *, int);

static void *
srealloc(void *vp, size_t nbytes)
{
	void *p;

	if ((p = realloc(vp, nbytes)) == NULL) {
		write(2, progname, strlen(progname));
		write(2, ": Insufficient memory.\n", 23);
		exit(077);
	}
	return p;
}

static void *
smalloc(size_t nbytes)
{
	return srealloc(NULL, nbytes);
}

static void
usage(void)
{
	fprintf(stderr, "usage: %s [-fir] file ...\n", progname);
	exit(2);
}

static void
msg(const char *fmt, ...)
{
	va_list	ap;

	if (!sysv3)
		fprintf(stderr, "%s: ", progname);
	va_start(ap, fmt);
	vfprintf(stderr, fmt, ap);
	va_end(ap);
}

static int
confirm(void)
{
	int yes = 0;
	char c;

	if (read(0, &c, 1) == 1) {
		yes = (c == 'y' || c == 'Y');
		while (c != '\n' && read(0, &c, 1) == 1);
	}
	return yes;
}

static void
setlevel(int level, struct getdb *db, int fd)
{
	if (level >= maxlevel)
		levels = srealloc(levels, (maxlevel=level+16) * sizeof *levels);
	levels[level].l_db = db;
	levels[level].l_fd = fd;
}

static size_t
catpath(size_t pend, const char *base)
{
	size_t	blen = strlen(base);

	if (pend + blen + 2 >= pathsz)
		path = srealloc(path, pathsz = pend + blen + 16);
	if (pend == 0 || path[pend-1] != '/')
		path[pend++] = '/';
	strcpy(&path[pend], base);
	return pend + blen;
}

static int
rmfile(const char *base, const struct stat *sp)
{
	if (iflag) {
		msg(sysv3 ? "%s ? " : "remove  %s: (y/n)? ", path);
		if (confirm() == 0)
			return 0;
	} else if ((sp->st_mode&S_IFMT) != S_IFDIR &&
			(sp->st_mode&S_IFMT) != S_IFLNK &&
			ontty && !fflag &&
			access(base, W_OK) < 0) {
		msg("%s: %o mode ? ", path, (int)(sp->st_mode & 0777));
		if (confirm() == 0)
			return 0;
	}
	if (((sp->st_mode&S_IFMT) == S_IFDIR ? rmdir : unlink)(base) < 0) {
#ifndef	SUS
		if ((sp->st_mode&S_IFMT) == S_IFDIR || !fflag || iflag)
#endif
			fprintf(stderr, (sp->st_mode&S_IFMT) == S_IFDIR ?
				"%s: Unable to remove directory %s\n%s\n" :
				"%s: %s not removed.\n%s\n",
				progname, path, strerror(errno));
		errcnt |= 1;
		return -1;
	}
	return 0;
}

static void
rm(size_t pend, const char *base, const int olddir, int ssub, int level)
{
	struct stat st;

	if (lstat(base, &st) < 0) {
		if (fflag == 0 || errno != ENOENT) {
			if (sysv3)
				fprintf(stderr, "%s: %s non-existent\n",
						progname, path);
			else
				fprintf(stderr, "%s: %s\n",
						path, strerror(errno));
			errcnt |= 4;
		}
		return;
	}
	if ((st.st_mode&S_IFMT) == S_IFDIR) {
		if (rflag) {
			struct direc *dp;
			struct getdb *db;
			int df, err;

			if (ssub == 0 && (iflag
#ifdef	SUS
					|| (!fflag && ontty &&
						access(base, W_OK) < 0)
#endif
					)) {
				msg("%s: directory %s: ? ", path);
				if (confirm() == 0)
					return;
			}
			if ((df = open(base,
							O_RDONLY
#ifdef	O_DIRECTORY
							| O_DIRECTORY
#endif
#ifdef	O_NOFOLLOW
							| O_NOFOLLOW
#endif
						)) < 0 ||
					(db = getdb_alloc(base, df)) == NULL) {
				if (errno == EMFILE) {
					int	sres;

					sres = subproc(pend, base, level);
					if (sres >= 0) {
						errcnt |= sres;
						goto remove;
					}
				}
				/*
				 * Maybe the directory is empty and can just
				 * be removed.
				 */
				if (rmfile(base, &st) < 0) {
					fprintf(stderr,
						"%s: cannot read "
						"directory %s\n",
						progname, path);
					errcnt |= 4;
				}
				return;
			}
			if (fchdir(df) < 0) {
				if (rmfile(base, &st) < 0) {
					fprintf(stderr,
						"%s: cannot chdir to %s\n",
							progname, path);
					errcnt |= 4;
				}
				getdb_free(db);
				close(df);
				return;
			}
			setlevel(level, db, df);
			while ((dp = getdir(db, &err)) != NULL) {
				if (dp->d_name[0] == '.' &&
						(dp->d_name[1] == '\0' ||
						 dp->d_name[1] == '.' &&
						 dp->d_name[2] == '\0'))
					continue;
				rm(catpath(pend, dp->d_name), dp->d_name,
						df, 0, level + 1);
				path[pend] = '\0';
			}
			if (err) {
				fprintf(stderr,
					"%s: error reading directory %s\n",
					progname, path);
				errcnt |= 4;
			}
			if (olddir >= 0 && fchdir(olddir) < 0) {
				fprintf(stderr, "%s: cannot change backwards\n",
						progname);
				exit(1);
			}
			getdb_free(db);
			close(df);
		} else {
			fprintf(stderr, "%s: %s directory\n", progname, path);
			errcnt |= 1;
			return;
		}
	}
	if (ssub == 0)
	remove:	rmfile(base, &st);
}

/*
 * POSIX.2 specifies that rm shall not fail because it exceeds the
 * maximum number of open file descriptors. It is not possible to
 * use path names and chdir("..") instead since the tree might be
 * moved during operation and we'd then start to remove files outside.
 * Creating a new process makes it possible to circumvent the OPEN_MAX
 * limit in a safe way.
 */
static int
subproc(size_t pend, const char *base, int level)
{
	pid_t	pid;

	switch (pid = fork()) {
	case 0: {
		int	i;

		errcnt = 0;
		for (i = 0; i < level - 1; i++) {
			getdb_free(levels[i].l_db);
			close(levels[i].l_fd);
		}
		rm(pend, base, -1, 1, 0);
		_exit(errcnt);
	}
	default: {
		int status;

		while (waitpid(pid, &status, 0) != pid);
		if (status && WIFSIGNALED(status)) {
			/*
			 * If the signal was sent due to a tty keypress,
			 * we should be terminated automatically and
			 * never reach this point. Otherwise, we terminate
			 * with the same signal, but make sure that we do
			 * not overwrite a possibly generated core file.
			 * This results in nearly the usual behavior except
			 * that the shell never prints a 'core dumped'
			 * message.
			 */
			struct rlimit	rl;

			rl.rlim_cur = rl.rlim_max = 0;
			setrlimit(RLIMIT_CORE, &rl);
			raise(WTERMSIG(status));
			pause();
		}
		return status ? WEXITSTATUS(status) : 0;
	}
	case -1:
		return -1;
	}
}

static char *
dotdot(const char *s)
{
	const char	*sp;

	for (sp = s; *sp; sp++);
	sp--;
	while (sp > s && *sp == '/')
		sp--;
	if (sp[0] == '.' && (sp == s || (sp > s && sp[-1] == '/')))
		return ".";
	if (&sp[-1] >= s && sp[0] == '.' && sp[-1] == '.' && (&sp[-1] == s ||
				&sp[-1] > s && sp[-2] == '/'))
		return "..";
	return NULL;
}

int
main(int argc, char **argv)
{
	int i, startfd = -1;

#ifdef	__GLIBC__
	putenv("POSIXLY_CORRECT=1");
#endif
	progname = basename(argv[0]);
	if (getenv("SYSV3") != NULL)
		sysv3 = 1;
	while ((i = getopt(argc, argv, "fiRr")) != EOF) {
		switch (i) {
		case 'f':
			fflag = 1;
#ifdef	SUS
			iflag = 0;
#endif
			break;
		case 'i':
#ifdef	SUS
			fflag = 0;
#endif
			iflag = 1;
			break;
		case 'R':
		case 'r':
			rflag = 1;
			break;
		default:
			usage();
		}
	}
#ifndef	SUS
	if (argv[optind] && argv[optind][0] == '-' && argv[optind][1] == '\0')
		optind++;
#endif
	if (optind >= argc)
		usage();
	ontty = isatty(0);
	if (rflag && (startfd = open(".", O_RDONLY)) < 0) {
		fprintf(stderr, "%s: cannot open current directory\n",
				progname);
		exit(1);
	}
	while (optind < argc) {
		if (dotdot(argv[optind]) != NULL) {
			if (fflag == 0)
				fprintf(stderr, "%s of %s is not allowed\n",
					progname, argv[optind]);
			errcnt |= 1;
		} else {
			if (path)
				free(path);
			i = strlen(argv[optind]);
			strcpy(path = smalloc(pathsz = i + 1), argv[optind]);
			rm(i, argv[optind], startfd, 0, 0);
		}
		optind++;
	}
	return errcnt;
}
